rfcs/text/0000-prefab-structs.md

```diff
@@ -0,0 +1,128 @@
+- Start Date: 2018-04-24
+- RFC PR: (leave this empty)
+- Feature Commit(s): (leave this empty)
+
+# Summary
+
+Typed Racket's handling of #:prefab structs is currently (as of April 2018) unsound.
+Instead of treating them like standard structs where we can assume (and enforce)
+that fields always contain values of a certain type (i.e. the current behabior), 
+prefabs should more-or-less be treated exactly like tagged n-ary cons/mcons cells.
+
+# Motivation
+
+Prefabs are clearly useful and desirable (e.g. they're serializable) and our current
+lack of sound support for them should be addressed.
+
+# Guide-level explanation
+
+## Prefab Top Types
+
+`(PrefabTop name)` is the top type for the prefab described by the
+identifier `name`. For immutable prefab structs, this type is equivalent
+(in fact short hand for) the type `(Prefab name Any ...)` (where there are
+as many occurrences of `Any` as there are fields in the prefab struct
+described by `name`). For mutable prefab structs, this type can be used for
+field reads (i.e. always producing a value of type `Any`) but cannot be used 
+for field writes (since the type of the invariant field is not known).
+
+
+## Immutable Prefabs
+
+This __immutable prefab defininition__:
+
+```racket
+(struct point ([x : Integer] [y : Integer])
+  #:prefab)
+```
+
+defines types/constructors/accessors for what can be thought of as a
+ _named tuple_ type, producing (more-or-less) the following type definition:
+
+`point ≈ (Prefab point Integer Integer)`
+
+and the following functions:
+
+- `point` of type `(-> Integer Integer point)`
+
+- `point?` of type `(-> Any Boolean : (Prefab point Any Any))` 
+(_NOTE: the type has `Any` not `Integer` in the field! This is because
+        a prefab could have been created anywhere with any value in
+        its fields_)
+
+- `point-x` of type `(All (X) (-> (Prefab Point X   Any) X))`
+
+- `point-y` of type `(All (Y) (-> (Prefab Point Any Y)   Y))`
+
+## Mutable Prefabs
+
+This __mutable prefab defininition__:
+
+```racket
+(struct initials ([first : Symbol] [last : Symbol])
+  #:prefab
+  #:mutable)
+```
+
+defines types/constructors/accessors for what can be thought of as 
+a _named heterogeneous vector_ type, producing (more-or-less) 
+the following type definition:
+
+`initials ≈ (Prefab initials Symbol Symbol)`
+
+and also produces the following definitions:
+
+- `initials` with type `(-> Symbols Symbols initials)`
+
+- `initials?` with type `(-> Any Boolean : (PrefabTop initials))`
+
+- `initials-first` with type `(All (F L) (-> (Prefab initials F L) F))` _and_
+type `(-> (PrefabTop initials) Any)` (i.e. a `case->`/intersection)
+
+- `initials-last` with type `(All (F L) (-> (Prefab initials F L) L))` _and_
+type `(-> (PrefabTop initials) Any)` (i.e. a `case->`/intersection)
+
+- `set-initials-first!` with type `(All (F L) (-> (Prefab initials F L) F Void))`
+
+- `set-initials-last!` with type `(All (F L) (-> (Prefab initials F L) L Void))`
```

Comment:
The `set!` functions should be `case->`s too, right? To work for the top type.

Reply:
No, I don't believe so. If we don't know what the mutable field types are, there's no safe way to write to them. Right?

Reply:
I'm not sure what you mean -- the type of `set-box!` doesn't mention `BoxTop`, and similarly for vectors.

Reply:
I don't see how these functions could have different types than `vector-set!` -- mutable prefabs are basically named vectors.